File: backend/src/main/java/com/hgw/gestionale/model/service/ModelService.java
package com.hgw.gestionale.model.service;

import com.hgw.gestionale.brand.entity.Brand;
import com.hgw.gestionale.brand.repository.BrandRepository;
import com.hgw.gestionale.common.exception.NotFoundException;
import com.hgw.gestionale.model.dto.CreateModelRequest;
import com.hgw.gestionale.model.dto.ModelResponse;
import com.hgw.gestionale.model.dto.UpdateModelRequest;
import com.hgw.gestionale.model.entity.Model;
import com.hgw.gestionale.model.mapper.ModelMapper;
import com.hgw.gestionale.model.repository.ModelRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class ModelService {

    private final ModelRepository modelRepository;
    private final BrandRepository brandRepository;

    public ModelResponse create(CreateModelRequest request) {
        Optional<Model> existing = modelRepository.findByNomeIgnoreCaseAndBrandNomeIgnoreCase(request.nome(), request.brandNome());
        if (existing.isPresent()) {
            return ModelMapper.toResponse(existing.get());
        }

        Brand brand = brandRepository.findByNomeIgnoreCase(request.brandNome())
                .orElseGet(() -> brandRepository.save(Brand.builder().nome(request.brandNome()).build()));

        return ModelMapper.toResponse(modelRepository.save(ModelMapper.toEntity(request, brand)));
    }

    public List<ModelResponse> getAll() {
        return modelRepository.findAll().stream()
                .map(ModelMapper::toResponse)
                .toList();
    }

    public List<ModelResponse> getByBrandId(Long brandId) {
        return modelRepository.findByBrandId(brandId).stream()
                .map(ModelMapper::toResponse)
                .toList();
    }

    public ModelResponse getById(Long id) {
        Model model = modelRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Model not found"));
        return ModelMapper.toResponse(model);
    }

    public ModelResponse getByName(String nome) {
        Model model = modelRepository.findByNomeIgnoreCase(nome)
                .orElseThrow(() -> new NotFoundException("Model not found"));
        return ModelMapper.toResponse(model);
    }

    public List<ModelResponse> search(String nome) {
        return modelRepository.findTop10ByNomeContainingIgnoreCase(nome)
                .stream()
                .map(ModelMapper::toResponse)
                .toList();
    }

    public List<ModelResponse> searchByBrand(Long brandId, String nome) {
        return modelRepository.findTop10ByBrandIdAndNomeContainingIgnoreCase(brandId, nome)
                .stream()
                .map(ModelMapper::toResponse)
                .toList();
    }

    public List<ModelResponse> searchByBrandName(String brandNome, String nome) {
        return modelRepository.findTop10ByBrandNomeContainingIgnoreCaseAndNomeContainingIgnoreCase(brandNome, nome)
                .stream()
                .map(ModelMapper::toResponse)
                .toList();
    }


    public ModelResponse update(Long id, UpdateModelRequest request) {
        Model model = modelRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Model not found"));

        Brand brand = brandRepository.findById(request.brandId())
                .orElseThrow(() -> new NotFoundException("Brand not found"));

        ModelMapper.updateEntity(model, request, brand);
        Model updated = modelRepository.save(model);

        return ModelMapper.toResponse(updated);
    }

    public void delete(Long id) {
        Model model = modelRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Model not found"));

        modelRepository.deleteById(model.getId());
    }
}

File: backend-node/src/service/model.service.ts
import { and, asc, count, desc, eq, ilike, SQL } from "drizzle-orm";
import { db } from "../config/db";
import { ModelResponse, CreateModelRequest, UpdateModelRequest, ModelSortBy } from "../dto/model.dto";
import { ModelMapper } from "../mapper/model.mapper";
import { models } from "../schema/models";
import { brands } from "../schema/brands";
import { HttpError } from "../common/httpError";
import { HttpStatus } from "../common/httpStatus";
import { PaginatedResponse, SortOrder } from "../common/pagination";

export class ModelService {

    async create(request: CreateModelRequest): Promise<ModelResponse> {
        const brandResult = await db.select().from(brands).where(ilike(brands.nome, request.brandNome));
        const brand = brandResult.at(0)
            ?? (await db.insert(brands).values({ nome: request.brandNome }).returning()).at(0)!;

        const [saved] = await db.insert(models).values({
            nome: request.nome,
            tipoDispositivo: request.tipoDispositivo,
            brandId: brand.id,
        }).returning();

        return ModelMapper.toResponse(saved, brand);
    }

    async getAllPaginated(
        page: number = 1,
        pageSize: number = 20,
        sortBy: ModelSortBy = "nome",
        sortOrder: SortOrder = "asc",
        nome?: string,
        brandNome?: string,
        tipoDispositivo?: string
    ): Promise<PaginatedResponse<ModelResponse>> {
        const colMap = { nome: models.nome, brand: brands.nome };
        const orderExpr = sortOrder === "desc" ? desc(colMap[sortBy]) : asc(colMap[sortBy]);

        const filters: SQL[] = [];
        if (nome) filters.push(ilike(models.nome, `%${nome}%`));
        if (brandNome) filters.push(ilike(brands.nome, `%${brandNome}%`));
        if (tipoDispositivo) filters.push(ilike(models.tipoDispositivo, tipoDispositivo));
        const where = filters.length > 0 ? and(...filters) : undefined;

        const [{ total }] = await db.select({ total: count() })
            .from(models)
            .innerJoin(brands, eq(models.brandId, brands.id))
            .where(where);

        const result = await db.select().from(models)
            .innerJoin(brands, eq(models.brandId, brands.id))
            .where(where)
            .orderBy(orderExpr)
            .limit(pageSize)
            .offset((page - 1) * pageSize);
        return {
            data: result.map(r => ModelMapper.toResponse(r.models, r.brands)),
            total,
            page,
            pageSize,
            totalPages: Math.ceil(total / pageSize),
        };
    }

    async getAll(): Promise<ModelResponse[]> {
        const result = await db.select().from(models).innerJoin(brands, eq(models.brandId, brands.id));
        return result.map(r => ModelMapper.toResponse(r.models, r.brands));
    }

    async search(nome: string): Promise<ModelResponse[]> {
        const result = await db.select().from(models)
            .innerJoin(brands, eq(models.brandId, brands.id))
            .where(ilike(models.nome, `%${nome}%`))
            .limit(10);
        return result.map(r => ModelMapper.toResponse(r.models, r.brands));
    }

    async searchByBrand(brandId: number, nome: string): Promise<ModelResponse[]> {
        const result = await db.select().from(models)
            .innerJoin(brands, eq(models.brandId, brands.id))
            .where(and(eq(models.brandId, brandId), ilike(models.nome, `%${nome}%`)))
            .limit(10);
        return result.map(r => ModelMapper.toResponse(r.models, r.brands));
    }

    async searchByBrandName(brandNome: string, nome: string): Promise<ModelResponse[]> {
        const result = await db.select().from(models)
            .innerJoin(brands, eq(models.brandId, brands.id))
            .where(and(ilike(brands.nome, `%${brandNome}%`), ilike(models.nome, `%${nome}%`)))
            .limit(10);
        return result.map(r => ModelMapper.toResponse(r.models, r.brands));
    }

    async getByBrandId(id: number): Promise<ModelResponse[]> {
        const result = await db.select().from(models)
            .innerJoin(brands, eq(models.brandId, brands.id))
            .where(eq(models.brandId, id));
        return result.map(r => ModelMapper.toResponse(r.models, r.brands));
    }

    async getByName(nome: string): Promise<ModelResponse> {
        const result = await db.select().from(models)
            .innerJoin(brands, eq(models.brandId, brands.id))
            .where(ilike(models.nome, nome));
        const row = result.at(0);
        if (!row) throw new HttpError(HttpStatus.NOT_FOUND, "Modello non trovato");
        return ModelMapper.toResponse(row.models, row.brands);
    }

    async getById(id: number): Promise<ModelResponse> {
        const result = await db.select().from(models)
            .innerJoin(brands, eq(models.brandId, brands.id))
            .where(eq(models.id, id));
        const row = result.at(0);
        if (!row) throw new HttpError(HttpStatus.NOT_FOUND, "Modello non trovato");
        return ModelMapper.toResponse(row.models, row.brands);
    }

    async update(id: number, request: UpdateModelRequest): Promise<ModelResponse> {
        const existing = await db.select().from(models).where(eq(models.id, id));
        if (!existing.at(0)) throw new HttpError(HttpStatus.NOT_FOUND, "Modello non trovato");

        const brandResult = await db.select().from(brands).where(eq(brands.id, request.brandId));
        const brand = brandResult.at(0);
        if (!brand) throw new HttpError(HttpStatus.NOT_FOUND, "Brand non trovato");

        const [updated] = await db.update(models).set({
            nome: request.nome,
            tipoDispositivo: request.tipoDispositivo,
            brandId: request.brandId,
        }).where(eq(models.id, id)).returning();

        return ModelMapper.toResponse(updated, brand);
    }

    async delete(id: number): Promise<void> {
        const result = await db.select().from(models).where(eq(models.id, id));
        if (!result.at(0)) throw new HttpError(HttpStatus.NOT_FOUND, "Modello non trovato");

        await db.delete(models).where(eq(models.id, id));
    }

}
